File: scrapers/inven.py
import re

import httpx

from scrapers.base import BaseScraper, Post

_MAN_RE = re.compile(r"([\d.]+)만")


class InvenScraper(BaseScraper):
    community = "inven"
    community_name = "인벤"
    base_url = "https://www.inven.co.kr"

    @staticmethod
    def _parse_views(text: str) -> int:
        """조회수 파싱 ('1.1만' → 11000, '5,559' → 5559)"""
        text = text.strip().replace(",", "")
        m = _MAN_RE.match(text)
        if m:
            return int(float(m.group(1)) * 10000)
        try:
            return int(text)
        except ValueError:
            return 0

    async def fetch_best_posts(self, client: httpx.AsyncClient) -> list[Post]:
        soup = await self._fetch_html(client, f"{self.base_url}/hot/")
        posts = []

        for item in soup.find_all("div", class_="list-common"):
            if "con" not in (item.get("class") or []):
                continue

            # 제목 링크
            title_div = item.find("div", class_="title")
            if not title_div:
                continue
            a_tag = title_div.find("a")
            if not a_tag:
                continue
            href = a_tag.get("href", "")
            if not href:
                continue

            # 제목: name에서 순번·카테고리 제거
            name_div = item.find("div", class_="name")
            if not name_div:
                continue
            title = name_div.get_text(strip=True)
            num_div = item.find("div", class_="num")
            cate_div = item.find("div", class_="cate")
            if num_div:
                num_text = num_div.get_text(strip=True)
                if title.startswith(num_text):
                    title = title[len(num_text):]
            if cate_div:
                cate_text = cate_div.get_text(strip=True)
                if title.startswith(cate_text):
                    title = title[len(cate_text):]
            title = title.strip()
            if not title:
                continue

            # 추천수
            votes = 0
            reco_div = item.find("div", class_="reco")
            if reco_div:
                try:
                    votes = int(reco_div.get_text(strip=True))
                except ValueError:
                    pass

            # 조회수
            views = 0
            hits_div = item.find("div", class_="hits")
            if hits_div:
                views = self._parse_views(hits_div.get_text(strip=True))

            # 댓글수
            comments = 0
            cmt_div = item.find("div", class_="comment")
            if cmt_div:
                m = re.search(r"\d+", cmt_div.get_text())
                if m:
                    comments = int(m.group())

            url = href if href.startswith("http") else f"{self.base_url}{href}"
            posts.append(self._make_post(title, url, votes=votes, views=views,
                                         comments=comments))

        return posts

File: internal/scraper/inven.go
package scraper

import (
	"fmt"
	"math"
	"net/http"
	"regexp"
	"strconv"
	"strings"

	"github.com/PuerkitoBio/goquery"
)

var (
	invenManRe    = regexp.MustCompile(`([\d.]+)만`)
	invenDigitRe  = regexp.MustCompile(`\d+`)
)

// InvenScraper scrapes 인벤 hot posts.
type InvenScraper struct {
	baseScraper
}

func NewInvenScraper() *InvenScraper {
	return &InvenScraper{
		baseScraper: baseScraper{
			community:     "inven",
			communityName: "인벤",
			baseURL:       "https://www.inven.co.kr",
			encoding:      "utf-8",
		},
	}
}

func (s *InvenScraper) Name() string { return s.communityName }

// parseInvenViews parses views like "1.1만" → 11000, "5,559" → 5559.
func parseInvenViews(text string) int {
	text = strings.TrimSpace(strings.ReplaceAll(text, ",", ""))
	m := invenManRe.FindStringSubmatch(text)
	if m != nil {
		f, err := strconv.ParseFloat(m[1], 64)
		if err == nil {
			return int(math.Round(f * 10000))
		}
	}
	v, err := strconv.Atoi(text)
	if err != nil {
		return 0
	}
	return v
}

func (s *InvenScraper) FetchBestPosts(client *http.Client) ([]Post, error) {
	doc, err := fetchDocument(client, fmt.Sprintf("%s/hot/", s.baseURL), s.encoding, s.baseURL)
	if err != nil {
		return nil, err
	}

	var posts []Post
	doc.Find("div.list-common").Each(func(_ int, item *goquery.Selection) {
		// Must have "con" class
		if classes, exists := item.Attr("class"); exists {
			if !strings.Contains(classes, "con") {
				return
			}
		} else {
			return
		}

		// Title link
		titleDiv := item.Find("div.title").First()
		if titleDiv.Length() == 0 {
			return
		}
		aTag := titleDiv.Find("a").First()
		if aTag.Length() == 0 {
			return
		}
		href, exists := aTag.Attr("href")
		if !exists || href == "" {
			return
		}

		// Title: from name div, strip num and cate prefixes
		nameDiv := item.Find("div.name").First()
		if nameDiv.Length() == 0 {
			return
		}
		title := strings.TrimSpace(nameDiv.Text())

		numDiv := item.Find("div.num").First()
		if numDiv.Length() > 0 {
			numText := strings.TrimSpace(numDiv.Text())
			if numText != "" && strings.HasPrefix(title, numText) {
				title = title[len(numText):]
			}
		}
		cateDiv := item.Find("div.cate").First()
		if cateDiv.Length() > 0 {
			cateText := strings.TrimSpace(cateDiv.Text())
			if cateText != "" && strings.HasPrefix(title, cateText) {
				title = title[len(cateText):]
			}
		}
		title = strings.TrimSpace(title)
		if title == "" {
			return
		}

		// Votes
		votes := 0
		recoDiv := item.Find("div.reco").First()
		if recoDiv.Length() > 0 {
			if v, err := strconv.Atoi(strings.TrimSpace(recoDiv.Text())); err == nil {
				votes = v
			}
		}

		// Views
		views := 0
		hitsDiv := item.Find("div.hits").First()
		if hitsDiv.Length() > 0 {
			views = parseInvenViews(hitsDiv.Text())
		}

		// Comments
		comments := 0
		cmtDiv := item.Find("div.comment").First()
		if cmtDiv.Length() > 0 {
			m := invenDigitRe.FindString(cmtDiv.Text())
			if m != "" {
				if v, err := strconv.Atoi(m); err == nil {
					comments = v
				}
			}
		}

		url := href
		if !strings.HasPrefix(href, "http") {
			url = s.baseURL + href
		}
		posts = append(posts, s.makePost(title, url, votes, views, comments))
	})

	return s.filterPosts(posts), nil
}
